lychee-lib/src/checker/website.rs

```diff
@@ -64,32 +74,77 @@ impl WebsiteChecker {
             retry_wait_time,
             accepted,
             require_https,
+            recursive_domains,
+            subsequent_uris: vec![],
         }
     }
 
     /// Retry requests up to `max_retries` times
     /// with an exponential backoff.
-    pub(crate) async fn retry_request(&self, request: Request) -> Status {
+    pub(crate) async fn retry_request(&self, request: Request) -> (Status, Vec<Uri>) {
         let mut retries: u64 = 0;
         let mut wait_time = self.retry_wait_time;
-        let mut status = self.check_default(clone_unwrap(&request)).await;
+        let (mut status, mut new_uris) = self.check_default(clone_unwrap(&request)).await;
         while retries < self.max_retries {
             if status.is_success() || !status.should_retry() {
-                return status;
+                return (status, new_uris);
             }
             retries += 1;
             tokio::time::sleep(wait_time).await;
             wait_time = wait_time.saturating_mul(2);
-            status = self.check_default(clone_unwrap(&request)).await;
+            (status, new_uris) = self.check_default(clone_unwrap(&request)).await;
         }
-        status
+        (status, vec![])
     }
 
     /// Check a URI using [reqwest](https://github.com/seanmonstar/reqwest).
-    async fn check_default(&self, request: Request) -> Status {
+    async fn check_default(&self, request: Request) -> (Status, Vec<Uri>) {
+        // println!("checking {}", request.url());
         match self.reqwest_client.execute(request).await {
-            Ok(ref response) => Status::new(response, self.accepted.clone()),
-            Err(e) => e.into(),
+            Ok(response) => {
+                let response_url = response.url().clone();
+                let status = Status::new(&response, self.accepted.clone());
+
+                if !self
+                    .recursive_domains
+                    .iter()
+                    .any(|d| response.url().domain() == Some(d))
+                {
+                    return (status, vec![]);
+                }
+
+                let mut base_url = response_url.clone();
+                base_url.set_path("/");
+
+                match response.text().await {
+                    Err(_) => (status, vec![]),
+                    Ok(response_text) => {
+                        let links: Vec<_> = Collector::new(None, Some(Base::Remote(base_url)))
```

Comment:
todo: use the collector that's constructed around `main`, since some flags are passed to it, while we are ouright ignoring a bunc of flags there